lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,416 @@
+    and moves them to the specified destination if either one of the following is true:
+
+    - The RSE expression of the rule is trivial (the RSE name itself).
+    - There are no replicas locked by the rule that reside on another RSE.
+
+    :param rse: RSE to decommission.
+    :param config: Decommissioning configuration dictionary.
+    :returns: A decommissioning profile dictionary.
+    """
+    try:
+        destination = config['destination']
+    except KeyError as exc:
+        raise RucioException('Configuration parameter "destination" not set') from exc
+
+    move_to_destination = RuleMover(destination)
+    return DecommissioningProfile(
+        rse=rse,
+        initializer=_set_availability,
+        discoverer=_generic_discover,
+        handlers=[
+            (_is_being_deleted, _count_as_processed),
+            (_has_child_rule_id, _count_as_processed),
+            (_has_trivial_rse_expression, move_to_destination),
+            (_has_last_replica_on_rse, move_to_destination),
+        ],
+        finalizer=_generic_finalize
+    )
+
+
+def _set_availability(
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log,
+) -> None:
+    """Initializer function that sets the RSE availability flags and the decommissioning status.
+
+    When an RSE is initialized for decommissioning, the availability flags are set as follows:
+
+    - `availability_read=True`
+    - `availability_write=False`
+    - `availability_delete=True`
+
+    and the following attribute is set:
+
+    - `greedyDeletion=True`
+
+    :param rse: RSE table entry as a dictionary.
+    :param logger: Logging function.
+    """
+    logger(logging.INFO,
+           '(%s) Setting RSE availability flags of to read !write delete, greedyDeletion=true.',
+           rse['rse'])
+
+    parameters = {
+        'availability_read': True,
+        'availability_write': False,
+        'availability_delete': True
+    }
+    update_rse(rse['id'], parameters)
+
+    try:
+        add_rse_attribute(rse['id'], 'greedyDeletion', True)
+    except Duplicate:
+        pass
+
+
+def _generic_discover(
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> Iterable[dict[str, Any]]:
+    """Discoverer function that calls the listing function from core.rule.
+
+    :param rse: RSE table entry as a dictionary.
+    :param logger: Logging function.
+    :returns: An iterable of rule dictionaries.
+    """
+    return list_rules_for_rse_decommissioning(rse['id'])
+
+
+def _generic_finalize(
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check for remaining replicas at the RSE and resolve where possible.
+
+    :param rse: RSE table entry as a dictionary.
+    :param logger: Logging function.
+    :returns: A boolean flag indicating the completion status.
+    """
+    replicas = list(list_replicas_per_rse(rse['id']))
```

Comment:
Is this always safe? Are we sure by this point that the amount of replicas returned is guaranteed to be low? Maybe there should be a limit, and the Decommissioner should rely on multiple calls of the finalizer function, if necessary, to complete the clean-up.

Reply:
I changed the way the remaining replicas are handled. Now we directly call `process_replicas_with_no_locks` with the generator from `list_replicas_per_rse` and let the function return the number of processed replicas, from which we decide whether the RSE is empty. Furthermore I set a limit of 100 replicas to process.

Reply:
I think there’s room for improvement here, but given that my immediate concern has been addressed, it’s fine to merge as is.